File: bindings/zenoh-c/src/zenoh_sub_thr.c
//
// Copyright (c) 2022 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//
#include <stdio.h>
#include <string.h>

#include "zenoh.h"


typedef struct {
    volatile unsigned long count;
    z_clock_t start;
    bool sleep;
} z_stats_t;

z_stats_t *z_stats_make() {
    z_stats_t *stats = z_malloc(sizeof(z_stats_t));
    stats->count = 0;
    stats->sleep = false;
    return stats;
}

void on_sample(const z_sample_t *sample, void *context) {
    z_stats_t *stats = (z_stats_t *)context;
    if (stats->sleep) {
        stats->count++;
    }
}
void drop_stats(void *context) {
    const z_stats_t *stats = (z_stats_t *)context;
    z_free(context);
}

struct args_t {
    unsigned int size;             // -s
    char* config_path;             // -c
    uint8_t help_requested;        // -h
};
struct args_t parse_args(int argc, char** argv);

int main(int argc, char **argv) {
    struct args_t args = parse_args(argc, argv);
    if (args.help_requested) {
        printf(
            "\
        -s (required, int): the size of the put message in bytes\n\
        -c (optional, string): the path to a configuration file for the session. If this option isn't passed, the default configuration will be used.\n\
        "
        );
        return 1;
    }

    z_owned_config_t config = args.config_path ? zc_config_from_file(args.config_path) : z_config_default();

    z_owned_session_t s = z_open(z_move(config));
    if (!z_check(s)) {
        printf("Unable to open session!\n");
        exit(-1);
    }

    z_owned_keyexpr_t ke = z_declare_keyexpr(z_loan(s), z_keyexpr("test/thr"));

    z_stats_t *context = z_stats_make();
    z_owned_closure_sample_t callback = z_closure(on_sample, drop_stats, context);
    z_owned_subscriber_t sub = z_declare_subscriber(z_loan(s), z_loan(ke), z_move(callback), NULL);
    if (!z_check(sub)) {
        printf("Unable to create subscriber.\n");
        exit(-1);
    }

    context->count = 0;
    context->sleep = true;
    context->start = z_clock_now();
    while (1) {
        z_sleep_s(1);
        if(context->count > 0){
            context->sleep = false;
            unsigned long elapsed = z_clock_elapsed_ms(&context->start);
            printf("%d,%.3f\n", args.size, (double)(context->count * 1000) / (double)(elapsed));
            fflush(stdout);
            // reset counter and timer
            context->count = 0;
            context->sleep = true;
            context->start = z_clock_now();
        }
    }

    z_undeclare_subscriber(z_move(sub));
    z_undeclare_keyexpr(z_loan(s), z_move(ke));
    z_close(z_move(s));
    return 0;
}

char* getopt(int argc, char** argv, char option) {
    for (int i = 0; i < argc; i++) {
        size_t len = strlen(argv[i]);
        if (len >= 2 && argv[i][0] == '-' && argv[i][1] == option) {
            if (len > 2 && argv[i][2] == '=') {
                return argv[i] + 3;
            } else if (i + 1 < argc) {
                return argv[i + 1];
            }
        }
    }
    return NULL;
}

struct args_t parse_args(int argc, char** argv) {
    for (int i = 0; i < argc; i++) {
        if (strcmp(argv[i], "-h") == 0) {
            return (struct args_t){.help_requested = 1};
        }
    }
    char* arg = getopt(argc, argv, 's');
    unsigned int size;
    if (arg) {
        size = atoi(arg);
    }
    else {
        return (struct args_t){.help_requested = 1};
    }
    return (struct args_t){.help_requested = 0,
                           .size = size,
                           .config_path = getopt(argc, argv, 'c')};
}

File: bindings/zenoh-python/src/zenoh_sub_thr.py
#
# Copyright (c) 2022 ZettaScale Technology
#
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
# which is available at https://www.apache.org/licenses/LICENSE-2.0.
#
# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
#
# Contributors:
#   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
#

import sys
import time
import argparse
import json
import zenoh
from zenoh import Reliability

# --- Command line argument parsing --- --- --- --- --- ---
parser = argparse.ArgumentParser(
    prog='z_sub_thr',
    description='zenoh throughput sub example')
parser.add_argument('--mode', '-m', dest='mode',
                    choices=['peer', 'client'],
                    type=str,
                    help='The zenoh session mode.')
parser.add_argument('--connect', '-e', dest='connect',
                    metavar='ENDPOINT',
                    action='append',
                    type=str,
                    help='Endpoints to connect to.')
parser.add_argument('--listen', '-l', dest='listen',
                    metavar='ENDPOINT',
                    action='append',
                    type=str,
                    help='Endpoints to listen on.')
parser.add_argument('--size', '-s', dest='payload_size',
                    metavar='PAYLOAD_SIZE',
                    required=True,
                    type=int,
                    help='Size of each put message in bytes.')
parser.add_argument('--config', '-c', dest='config',
                    metavar='FILE',
                    type=str,
                    help='A configuration file.')

args = parser.parse_args()
conf = zenoh.Config.from_file(args.config) if args.config is not None else zenoh.Config()
if args.mode is not None:
    conf.insert_json5(zenoh.config.MODE_KEY, json.dumps(args.mode))
if args.connect is not None:
    conf.insert_json5(zenoh.config.CONNECT_KEY, json.dumps(args.connect))
if args.listen is not None:
    conf.insert_json5(zenoh.config.LISTEN_KEY, json.dumps(args.listen))
size = args.payload_size

count = 0
start = None
sleeping = False

def listener(sample):
    global count, sleeping
    if sleeping:
        count += 1

def main():
    global count, start, sleeping
    # initiate logging
    zenoh.init_logger()

    session = zenoh.open(conf)

    # By explicitly constructing the `Closure`, the `Queue` that's normally inserted between the callback and zenoh is removed.
    # Only do this if your callback runs faster than the minimum expected delay between two samples.
    sub = session.declare_subscriber("test/thr", zenoh.Closure((listener, lambda *args: None)), reliability=Reliability.RELIABLE())

    count = 0
    sleeping = True
    start = time.time()
    while True:
        time.sleep(1)
        if count > 0:
            sleeping = False
            end=time.time()
            print(f"{size},{(count / (end - start)):.3f}", flush=True)
            # reset counter and timer
            count = 0
            sleeping = True
            start = time.time()

    sub.undeclare()
    session.close()
    # while `sub.undeclare()` only returns once the unsubscription is done (no more callbacks will be queued from that instant), already queued callbacks may still be running in threads that Python can't see.
    time.sleep(0.1)

main()
